### PR TITLE
[Core][C++ Worker] libray_api.so exports absl::flag symbols

### PR DESCRIPTION
**Pull request by larrylian**
## Why are these changes needed?
1. If user driver also use absl::flag when link libray_api.so,  Then when the driver is running, fill in the ray related parameters(ray_**), it will report an error "ERROR: Unknown command line flag" 

2. Remove unuse file "ray_api_exported_symbols_mac.lds"

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: cpp/symbols/ray_api_exported_symbols_linux.lds [hunk 1]
```diff
@@ -10,6 +10,10 @@ VERSION_1.0 {
         GetFunctionManager;
         GetRemoteFunctions;
         InitRayRuntime;
+        # If the symbol of absl::flags is hidden, and the user both use the absl::flags library,
+        # then the flags feature will have problems.
```
Comment (SongGuyang) [inline cpp/symbols/ray_api_exported_symbols_linux.lds:14]:
```suggestion
        #  If the symbols of absl::flags are hidden, and the users also use the absl::flags library, then the flags feature will bring issues, e.g. `ERROR: Unknown command line flag`.
```

**Review by SongGuyang — COMMENTED** [review]
Can we add a test for it?

**Comment by larrylian** [comment]
> Can we add a test for it?

The current UT driver is statically connected to libray_api.a. Later, I will reconstruct UT and let the driver dynamically connect libray_api.so to complete this UT.

**Review by SongGuyang — APPROVED** [review]
LGTM

**Comment by ArturNiederfahrenhorst** [comment]
Ran into issues compiling on an M1. Might be related to this PR:

```
ERROR: file 'cpp/libray_api.so' is generated by these conflicting actions:
    Label: //cpp:libray_api.so, //cpp:ray_api
    RuleClass: cc_binary rule, cc_library rule
    Configuration: 67199d4deec70d3c872536c8409a4659ddffa610cdcd97874ca1b2d2fc20f2a2
    Mnemonic: CppLink, Fail
    Action key: a91e191a2ee12a610beebda015663e146f9765a96e1efd4b7338fee3ce53d500, 084cdee4396a3c44f00642bce29f9d7119113866a06d29929bd6e1f6c78fdbab
    Progress message: Linking cpp/libray_api.so, Reporting failed target //cpp:ray_api located at /Users/artur/code/ray/cpp/BUILD.bazel:30:11
    PrimaryInput: File:[[<execution_root>]bazel-out/darwin_arm64-opt/bin]cpp/libray_api.lo, (null)
    PrimaryOutput: File:[[<execution_root>]bazel-out/darwin_arm64-opt/bin]cpp/libray_api.so
    ERROR: com.google.devtools.build.lib.skyframe.ArtifactConflictFinder$ConflictException: com.google.devtools.build.lib.actions.MutableActionGraph$ActionConflictException: for cpp/libray_api.so, previous action: action 'Reporting failed target //cpp:ray_api located at /Users/artur/code/ray/cpp/BUILD.bazel:30:11', attempted action: action 'Linking cpp/libray_api.so'

```

**Comment by SongGuyang** [comment]
@ArturNiederfahrenhorst Thanks for your report. @larrylian Take a look this?

**Comment by larrylian** [comment]
@ArturNiederfahrenhorst  Can you provide more information? For example, to reproduce the command. According to your error log, it is likely that there is a problem with your own bazel compilation dependency configuration.

**Comment by ArturNiederfahrenhorst** [comment]
I was not sure if it was my setup or this change that broke things so I just left this breadcrumb here instead anyone else would stumble across it.

**Comment by ArturNiederfahrenhorst** [comment]
@scottjlee Just ran into the same issue and we'll work together to provide the info you need I guess.

**Comment by scottjlee** [comment]
I ran the command `pip install -e . --verbose` in the `python/` directory (following [build steps](https://docs.ray.io/en/latest/ray-contribute/development.html#building-ray-on-linux-macos-full)).
The output I get is:

```
ERROR: file 'cpp/libray_api.so' is generated by these conflicting actions:
    Label: //cpp:libray_api.so, //cpp:ray_api
    RuleClass: cc_binary rule, cc_library rule
    Configuration: 67199d4deec70d3c872536c8409a4659ddffa610cdcd97874ca1b2d2fc20f2a2
    Mnemonic: CppLink, Fail
    Action key: a91e191a2ee12a610beebda015663e146f9765a96e1efd4b7338fee3ce53d500, 084cdee4396a3c44f00642bce29f9d7119113866a06d29929bd6e1f6c78fdbab
    Progress message: Linking cpp/libray_api.so, Reporting failed target //cpp:ray_api located at /Users/sjl/src/ray/cpp/BUILD.bazel:29:11
    PrimaryInput: File:[[<execution_root>]bazel-out/darwin_arm64-opt/bin]cpp/libray_api.lo, (null)
    PrimaryOutput: File:[[<execution_root>]bazel-out/darwin_arm64-opt/bin]cpp/libray_api.so
    ERROR: com.google.devtools.build.lib.skyframe.ArtifactConflictFinder$ConflictException: com.google.devtools.build.lib.actions.MutableActionGraph$ActionConflictException: for cpp/libray_api.so, previous action: action 'Reporting failed target //cpp:ray_api located at /Users/sjl/src/ray/cpp/BUILD.bazel:29:11', attempted action: action 'Linking cpp/libray_api.so'
    INFO: Elapsed time: 39.708s
    INFO: 0 processes.
    FAILED: Build did NOT complete successfully (173 packages loaded, 17162 targets configured)
```
The Bazel version is 5.4.0, I am running this on an M1 Mac.

**Comment by pcmoritz** [comment]
This is the hack I used to fix it, I'm glad I'm not the only one with this problem:
```
diff --git a/cpp/BUILD.bazel b/cpp/BUILD.bazel
index 52db9f78bc..8e33af0653 100644
--- a/cpp/BUILD.bazel
+++ b/cpp/BUILD.bazel
@@ -21,13 +21,13 @@ cc_binary(
     linkstatic = 1,
     visibility = ["//visibility:public"],
     deps = [
-        ":ray_api",
+        ":the_ray_api",
         ":symbols/ray_api_exported_symbols_linux.lds",
     ],
 )
 
 cc_library(
-    name = "ray_api",
+    name = "the_ray_api",
     srcs = glob([
         "src/ray/api.cc",
         "src/ray/api/*.cc",
@@ -95,7 +95,7 @@ cc_binary(
         "@bazel_tools//src/conditions:windows": [
             # TODO(SongGuyang): Change to use dynamic library
             # "ray_cpp_lib" when we make it work on Windows.
-            "ray_api",
+            "the_ray_api",
         ],
         "//conditions:default": [
             "ray_cpp_lib",
@@ -111,7 +111,7 @@ genrule(
     name = "ray_cpp_pkg",
     srcs = [
         "default_worker",
-        "ray_api",
+        "the_ray_api",
         "libray_api.so",
     ],
     outs = ["ray_cpp_pkg.out"],
@@ -175,7 +175,7 @@ cc_test(
     linkstatic = True,
     tags = ["team:core"],
     deps = [
-        "ray_api",
+        "the_ray_api",
         "@com_google_googletest//:gtest_main",
     ],
 )
@@ -205,7 +205,7 @@ cc_test(
     linkstatic = True,
     tags = ["team:core"],
     deps = [
-        "ray_api",
+        "the_ray_api",
         "@com_google_googletest//:gtest_main",
     ],
 )
@@ -228,7 +228,7 @@ cc_test(
     linkstatic = True,
     tags = ["team:core"],
     deps = [
-        "ray_api",
+        "the_ray_api",
         "@com_google_googletest//:gtest_main",
     ],
 )
@@ -293,7 +293,7 @@ cc_test(
     linkstatic = True,
     tags = ["team:core"],
     deps = [
-        "ray_api",
+        "the_ray_api",
     ],
 )
 
@@ -327,7 +327,7 @@ cc_binary(
     linkstatic = True,
     tags = ["team:core"],
     deps = [
-        ":ray_api",
+        ":the_ray_api",
     ],
 )
 

```

I didn't have time to test this or do more with this, let's open an issue and investigate?

**Comment by larrylian** [comment]
@pcmoritz @scottjlee  
Thanks for your feedback. I will try to solve the problem as soon as possible within today. I guess it may have something to do with the bazel version upgrade. Cause semantic conflict between ray_api and libray_api.so

**Pull request merged**